src/librustc_mir/transform/preallocate_gc_contents.rs

```diff
@@ -10,47 +10,72 @@ use rustc_span::DUMMY_SP;
         let param_env = tcx.param_env(source.def_id()).with_reveal_all();
-        let mut patch = MirPatch::new(body);
 
+        GcPatcher { tcx, source, param_env }.run_pass(body);
+    }
+}
+
+struct GcPatcher<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    source: MirSource<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+}
+
+impl GcPatcher<'tcx> {
+    fn run_pass(&mut self, body: &mut BodyAndCache<'tcx>) {
+        let mut patch = MirPatch::new(body);
         for (bb, bb_data) in body.basic_blocks().iter_enumerated() {
             let terminator = bb_data.terminator();
             if let TerminatorKind::Call {
-                func: ref op,
+                ref func,
                 ref args,
                 ref destination,
                 cleanup,
                 from_hir_call,
             } = terminator.kind
             {
-                if let ty::FnDef(callee_def_id, substs) = op.ty(&**body, tcx).kind {
-                    let name = tcx.def_path_str(callee_def_id);
-                    if name == VEC_CONSTRUCTOR {
-                        let elem_ty = substs.type_at(0);
-                        let impls_manageable_contents =
-                            ty_impls_manageable_contents(elem_ty, tcx, param_env);
-                        if impls_manageable_contents {
-                            // If our T in Vec<T> implements ManageableContents, then we patch
-                            // Vec::new to use BoehmAllocator instead of Global alloc.
-                            let new_in_col_fn = tcx.lang_items().new_in_collector_fn().unwrap();
-                            let new_in_col_fn_op =
-                                Operand::function_handle(tcx, new_in_col_fn, substs, DUMMY_SP);
-                            let new_term = TerminatorKind::Call {
-                                func: new_in_col_fn_op,
+                if let ty::FnDef(callee_def_id, substs) = func.ty(&**body, self.tcx).kind {
+                    let name = self.tcx.def_path_str(callee_def_id);
+                    let new_term = match name.as_str() {
+                        VEC_WITH_CAPACITY => {
+                            if !ty_impls_manageable_contents(substs.type_at(0), self.tcx, self.param_env) {
+                                continue
+                            }
+                            let new_callee = self.tcx.lang_items().vec_with_capacity_fn().unwrap();
+                            let func = Operand::function_handle(self.tcx, new_callee, substs, DUMMY_SP);
+                            TerminatorKind::Call {
+                                func,
+                                args: args.clone(),
+                                destination: destination.clone(),
+                                cleanup,
+                                from_hir_call
+                            }
+                        },
+                        VEC_PUSH => {
```

Comment:
Not that I think it matters for this PR, but does this overriding of `push` imply that if one calls (say) `Vec::insert` that we won't use the GC'd method for extending the `Vec` and something bad will happen?

Reply:
Yes. Unfortunately this was quite a lot more difficult than I'd anticipated because it's impossible to build a `Vec` with a custom backing allocator. At the moment I've got it working for the common case only.

Reply:
That's fine for this PR, so long as we know explicitly that's a limitation.

Reply:
One thing to note though is that this patching is only necessary for the first allocation in the vector. E.g. If I do `vec.push()` and then `vec.insert()`, the first patched allocation will mean that any `realloc` calls are also put in the GC heap because Boehm looks at where the item was previously allocated.